higher-order-functions/writing-higher-order-functions.md

```diff
@@ -0,0 +1,171 @@
+
+```py    
+def get_first_movie(movies):
+    min_movie = movies[0]
+    min_title = movies[0]["title"]
+    for movie in movies:
+        if movie["title"] < min_title:
+            min_movie = movie
+            min_title = movie["title"]
+    return min_movie
+
+def get_earliest_movie(movies):
+    min_movie = movies[0]
+    min_date = movies[0]["release_year"]
+    for movie in movies:
+        if movie["release_year"] < min_date:
+            min_movie = movie
+            min_date = movie["release_year"]
+    return min_movie
+
+def get_lowest_rated_movie(movies):
+    min_movie = movies[0]
+    min_rating = movies[0]["rating"]
+    for movie in movies:
+        if movie["rating"] < min_rating:
+            min_movie = movie
+            min_rating = movie["rating"]
+    return min_movie
+```
+
+We then saw how we could DRY up this code with the `min` function and the optional `key` parameter.
+
+```py
+def get_title(movie):
+    return movie["title"]
+
+
+def get_rating(movie):
+    return movie["rating"]
+
+
+def get_release_year(movie):
+    return movie["release_year"]
+
+
+first_movie_by_title = min(movies, key=get_title)
+lowest_rated_movie = min(movies, key=get_rating)
+earliest_movie = min(movies, key=get_release_year)
+```
+
+Finally, in the lesson on [lambda expressions](./lambdas.md), we achieved the functionality above with more concise code:
+
+```py
+first_movie_by_title = min(movies, key=lambda movie:movie["title"])
+lowest_rated_movie = min(movies, key=lambda movie:movie["rating"])
+earliest_movie = min(movies, key=lambda movie:movie["release_year"])
+```
+
+## Writing a Custom Min Function
+
+Now, let's demystify the `min` function with the optional `key` parameter by writing our own custom min function `min_function_custom`. We will use `get_first_movie` as a reference:
+
+```py
+def get_first_movie(movies):
+    min_movie = movies[0]
+    min_title = movies[0]["title"]
+    for movie in movies:
+        if movie["title"] < min_title:
+            min_movie = movie
+            min_title = movie["title"]
+    return min_movie
+```
+
+Notice that we can replace each line with the expression `movie["title"]` with a function call `get_title(movies[0])` without changing the behavior of the function. 
+
+```py
+def get_title(movie):
+    return movie["title"]
+
+
+def get_first_movie(movies):
+    min_movie = movies[0]
+    min_title = get_title(movies[0])
+    for movie in movies:
+        if get_title(movie) < min_title:
+            min_movie = movie
+            min_title = get_title(movie)
+    return min_movie
+```
+
+Now, let's take it one step further to generalize our function to find the minimum movie by any key (i.e. `"title"`, `"rating"`, or `"release_year"`). 
```

Comment:
We talk about finding the minimum of a movie based on attributes on this line, but the the description of the step removes all movie-related naming. I'm wondering if it's worth splitting out the renaming as a final step - something like "We now have a flexible custom min function for movies, but this would work for any list of dictionaries. Let's rename our variables for clarity!"